library/core/src/ptr/mod.rs

```diff
@@ -427,6 +427,237 @@ pub use unique::Unique;
 mod const_ptr;
 mod mut_ptr;
 
+/// The list of address spaces.
+///
+/// See [`AddrspacePtr`] for using pointers with an explicit associated address space.
+///
+/// This module lists the address spaces for the current target.
+#[unstable(feature = "ptr_addrspace", issue = "none")]
+pub mod addrspace {
+    /// The default address space.
+    ///
+    /// On most targets, a raw pointer (`*mut/const T`) implicitly points into the generic address space.
+    #[unstable(feature = "ptr_addrspace", issue = "none")]
+    pub const GENERIC: u32 = 0;
+
+    /// The address space corresponding to VRAM on GPUs.
+    ///
+    /// It is readable and writable from the GPU and CPU.
+    /// Pointers into this address space can be casted to and from the generic address space.
+    #[unstable(feature = "ptr_addrspace", issue = "none")]
+    #[cfg(any(doc, target_arch = "amdgpu", target_arch = "nvptx64"))]
+    #[doc(cfg(any(target_arch = "amdgpu", target_arch = "nvptx64")))]
+    pub const GLOBAL: u32 = 1;
+
+    /// The address space shared between threads within the same workgroup on GPUs.
+    ///
+    /// It is readable and writable from the GPU.
+    /// Pointers into this address space can be casted to and from the generic address space.
+    #[unstable(feature = "ptr_addrspace", issue = "none")]
+    #[cfg(any(doc, target_arch = "amdgpu", target_arch = "nvptx64"))]
+    #[doc(cfg(any(target_arch = "amdgpu", target_arch = "nvptx64")))]
+    pub const WORKGROUP: u32 = 3;
+
+    /// The address space for constant memory on GPUs.
+    ///
+    /// It is readable from the GPU but not writable.
+    /// Pointers into this address space can be casted to and from the generic address space.
+    #[unstable(feature = "ptr_addrspace", issue = "none")]
+    #[cfg(any(doc, target_arch = "amdgpu", target_arch = "nvptx64"))]
+    #[doc(cfg(any(target_arch = "amdgpu", target_arch = "nvptx64")))]
+    pub const CONST: u32 = 4;
+}
+
+/// A pointer into a target-specific address space.
+///
+/// An address space is a special, target-specific memory region.
+/// Each pointer has an associated address space. If no address space is explicitly mentioned, it
+/// is usually the generic address space. When accessing low-level hardware capabilities or
+/// intrinsics, it is sometimes necessary to explicitly specify a different address space.
+/// For these cases, `AddrspacePointer` can be used.
+///
+/// See also the [pointer](pointer) primitive types and the [`addrspace`] module.
+#[unstable(feature = "ptr_addrspace", issue = "none")]
+#[lang = "addrspace_ptr_type"]
+#[allow(missing_debug_implementations)]
+pub struct AddrspacePtr<T: 'static, const ADDRSPACE: u32> {
+    // Struct implementation is replaced by the compiler.
+    // This field is here for using the generic argument but cannot be set or accessed in any way.
+    do_not_use: PhantomData<*const T>,
```

Comment:
This seems like a very fragile hack -- having fields that then get removed half-way through compilation. There's a huge risk of confusing some part of the compiler that still looks at those fields. Not even `Box` does something this cursed, and that type is already quite cursed...